### DISCUSSION TITLE
Max Size of a Firecracker VM

### DISCUSSION THREAD

**Discussion by SortaASoraous** [firecracker-microvm/firecracker#3092] [Q&A]
As the title states, what is the maximum "size" (in terms of CPU, memory, bandwidth, and other relevant compute resources) that I can reasonably allocate to a VM?

**Comment by dianpopa** [comment]
HI!

The maximum number of CPUs a microVM can have is [32](https://github.com/firecracker-microvm/firecracker/blob/196ac5a4c38adb9dcd8068c02c93c2a22ba8b726/src/vmm/src/vmm_config/machine_config.rs#L12).

As far as memory limit, we do not have one inside Firecracker, however we depend on KVM limits. KVM supports memory slots up to [KVM_MEM_MAX_NR_PAGES](https://elixir.free-electrons.com/linux/v4.14/source/include/linux/kvm_host.h#L294) pages.

As far as network, you can take a look at https://github.com/firecracker-microvm/firecracker/blob/main/docs/network-performance.md to form an idea of the performance numbers.

**Comment by dianpopa** [comment]
@SortaASoraous does this answer your enquiry?

**Reply by majek** [reply]
> The maximum number of CPUs a microVM can have is [32](https://github.com/firecracker-microvm/firecracker/blob/196ac5a4c38adb9dcd8068c02c93c2a22ba8b726/src/vmm/src/vmm_config/machine_config.rs#L12).

Ok, but why? Can I change this constant?

**Reply by xmarcalx** [reply]
Hi @majek ,

As pointed in a slack conversation in our [firecracker channel](https://firecracker-microvm.slack.com/archives/CDL3FUR8B/p1684174039806349) theoretical there is no limit to the amount of vCPU that a microVM can support.
However, empirically we observed that increasing number of vCPU in the microVMs does not correspond to a real increase in performance of the microVM itself because of the contention between those vCPU threads and the single VMM thread which need to handle messages for all those vCPU threads and the devices.

You can simplytest by yourself, creating a microVM with for example 8,16, 32 and 64 vCPUs, make sure that each thread is busy and plot the performance of each microVM and you will be able to see a knee point in the curve.
At the same time we do not monitor this KPI because so far we never had the usecase of microVMs with more than 32 vCPUs.

I hope this answer your question.

**Comment by ShadowCurse** [comment]
Closing for now. If anyone wants more details please reopen.
